dot/rpc/modules/state.go

```diff
@@ -489,6 +495,41 @@ func (sm *StateModule) QueryStorage(
 	return nil
 }
 
+// QueryStorageAt queries historical storage entries (by key) at the best block if the given block hash is nil
+func (sm *StateModule) QueryStorageAt(
+	_ *http.Request, req *StateStorageQueryAtRequest, res *[]StorageChangeSetResponse) error {
+	var atBlockHash common.Hash
+	if req.At.IsEmpty() {
+		atBlockHash = sm.blockAPI.BestBlockHash()
+	} else {
+		atBlockHash = req.At
+	}
+
+	response := make([]StorageChangeSetResponse, 0, 1)
+	changes := make([][2]*string, 0, len(req.Keys))
+
+	for _, key := range req.Keys {
+		value, err := sm.storageAPI.GetStorageByBlockHash(&atBlockHash, common.MustHexToBytes(key))
+		if err != nil {
+			return fmt.Errorf("getting value by block hash: %w", err)
+		}
+		var hexValue *string
+		if len(value) > 0 {
+			hexValue = stringPtr(common.BytesToHex(value))
+		}
+
+		changes = append(changes, [2]*string{stringPtr(key), hexValue})
+	}
+
+	response = append(response, StorageChangeSetResponse{
+		Block:   &atBlockHash,
+		Changes: changes,
+	})
+
+	*res = response
```

Comment:
nit change to directly be:

```go
	res = new([]StorageChangeSetResponse)
	*res = []StorageChangeSetResponse{{
		Block:   &atBlockHash,
		Changes: changes,
	}}
```

that way you can also remove the init `response := make([]StorageChangeSetResponse, 0, 1)` above and avoid name conflicts between `res` and `response`.